File: backend/services/membership-service/internal/dto/membership.go
package dto

import "time"

// MembershipResponse 是一条会员资格的对外形状。
//
// 它同时服务后台详情页、后台列表、小程序会员中心三处。三处要看的字段有重叠也有差别（后台要
// 看冻结原因与操作留痕，小程序不看），但**形状是同一个**：分两套结构会让「后台显示还在有效期、
// 小程序显示已过期」这种不一致有地方发生，而它正是最不该有的那类 bug。
//
// 列表与详情也因此同形，区别只在填了哪些部分（详情多一个 Changes）。
type MembershipResponse struct {
	ID     string `json:"id"`
	UserID string `json:"userId"`
	// PlanCode / PlanName 是**成交当时的快照**，不是现查套餐。套餐后来改了名，这里不跟着变
	// ——用户当初买的就是那个名字。
	PlanCode string `json:"planCode"`
	PlanName string `json:"planName"`
	// MemberPriceMode 同样是成交快照：auto=会员本人自动享会员价，coupon=靠会员价体验券。
	// 它是「这个人现在怎么才能享到会员价」的唯一答案，前端据此决定说哪句话。
	MemberPriceMode string `json:"memberPriceMode"`
	// StoreID / StoreName 是**归属门店**：这个人是谁拉来的。
	//
	// 它不参与权益判定、不影响核销、不参与分账——后台把它展示在用户详情与会员详情上，仅此
	// 而已。StoreName 是**商户域的事实**，本库不留第二份：每次读一页由服务层现解一次
	// （见 service.StoreNames），解不出来时是空串而不报错。
	StoreID   string `json:"storeId"`
	StoreName string `json:"storeName"`
	// active / frozen / expired / revoked。
	Status   string    `json:"status"`
	StartAt  time.Time `json:"startAt"`
	ExpireAt time.Time `json:"expireAt"`
	// Active 是**算出来给前端用的**：这个会员此刻到底还作不作数（status=active 且在有效期内）。
	//
	// 它不进库、不改任何东西，纯粹是把「status 与 expire_at 一起看」这条判断从三处前端里
	// 收回来一处。到期扫描没有跑完的那一小段时间里，status 还是 active 而 expire_at 已经过了
	// ——只看 status 的前端会在那一小段时间里显示「会员有效」。
	Active bool `json:"active"`
	// AutoRenew 是用户可关的那个开关。套餐支不支持自动续费看 PlanResponse.AutoRenew，
	// 两者不是一回事：套餐说「这个产品可以签代扣」，这里说「这个人现在开着」。
	AutoRenew      bool       `json:"autoRenew"`
	AutoRenewOffAt *time.Time `json:"autoRenewOffAt"`
	RenewalCount   int32      `json:"renewalCount"`
	LastRenewedAt  *time.Time `json:"lastRenewedAt"`
	// 冻结与撤销的留痕。Reason 是系统给的短因由，Remark 是后台人工填的备注。
	FrozenAt     *time.Time `json:"frozenAt"`
	FreezeReason string     `json:"freezeReason"`
	RevokedAt    *time.Time `json:"revokedAt"`
	RevokeReason string     `json:"revokeReason"`
	CreatedAt    time.Time  `json:"createdAt"`
	UpdatedAt    time.Time  `json:"updatedAt"`
	// Changes 是时间线，**只有详情填**。列表页空数组：一页 20 条各自的完整流水进不了列表，
	// 硬塞进去只会让列表接口变成一个没人敢改的重查询。
	Changes []ChangeResponse `json:"changes"`
}

// ChangeResponse 是会员身上发生过的一件事。
type ChangeResponse struct {
	ID string `json:"id"`
	// 取值见 model.Change*：activate / renew / expire / freeze / unfreeze / auto_renew_on /
	// auto_renew_off / subscribe / unsubscribe / refund_adjust / admin_adjust / revoke。
	//
	// 回的是**编码**而不是中文：文案归前端（它有标签表），而且同一件事在不同页面上的说法
	// 可以不一样（「开通」在时间线上、在筛选下拉里都合适，在按钮上就不合适）。
	ChangeType string `json:"changeType"`
	// From / To 两对，可能为空：续费只动有效期（状态两个都是 active），冻结只动状态。
	FromStatus   string     `json:"fromStatus"`
	ToStatus     string     `json:"toStatus"`
	FromExpireAt *time.Time `json:"fromExpireAt"`
	ToExpireAt   *time.Time `json:"toExpireAt"`
	// OrderID 为空是常态里的常态：到期扫描、后台调整都没有订单。
	OrderID *string `json:"orderId"`
	// user / admin / system / worker。
	OperatorType string    `json:"operatorType"`
	OperatorID   *string   `json:"operatorId"`
	Reason       string    `json:"reason"`
	Remark       string    `json:"remark"`
	OccurredAt   time.Time `json:"occurredAt"`
}

// MiniappMembershipResponse 是小程序「我的会员」那一屏的响应。
//
// 包一层而不是直接回 MembershipResponse，是为了让**「还不是会员」也是一个 200**：
// 绝大多数用户打开会员中心时都不是会员，把它做成 404 会让前端每一次正常访问都走错误分支，
// 而错误分支里多半会弹一个「加载失败」。包一层之后，null 是明确的、可判定的答案。
//
// 它**不复用后台的 MembershipResponse**：后台那个带着冻结原因、撤销原因、操作留痕——那些是
// 内部管理信息，不该出现在用户的手机上（与 dto.MiniappPlanResponse 同一条理由）。
type MiniappMembershipResponse struct {
	Membership *MiniappMembership `json:"membership"`
}

// MiniappMembership 是用户自己看到的那份会员。
//
// 与后台形状的差别不只是字段多少：这里没有 user_id（用户当然知道是自己），没有冻结与撤销的
// 原因（那是运营写给自己人看的），也没有时间线（那是后台对账用的）。
type MiniappMembership struct {
	ID       string `json:"id"`
	PlanName string `json:"planName"`
	// MemberPriceMode 决定这一屏上说哪句话：「你已享会员价」还是「你有 N 张会员价券」。
	MemberPriceMode string    `json:"memberPriceMode"`
	Status          string    `json:"status"`
	StartAt         time.Time `json:"startAt"`
	ExpireAt        time.Time `json:"expireAt"`
	// Active 是算出来的（status=active 且在有效期内），与后台那个同名。
	Active    bool `json:"active"`
	AutoRenew bool `json:"autoRenew"`
	// MemberPriceCouponsPerPeriod 是**成交快照**上「每期发几张」，coupon 模式下用来显示
	// 「本月可享 20 次会员价」。它不是余额——券在 coupon-service，余额要在那里查。
	//
	// **没有「这个套餐支不支持自动续费」这个字段**：memberships 上没有它，而现查套餐（按
	// plan_id）与「成交快照说了算」这条本域反复申明的规矩直接冲突——套餐今天关掉自动续费，
	// 不等于这个人当初买的那个产品不支持。凡是需要它的地方（购买页、续费入口）本来就有套餐
	// 列表，那里是它的家。
	MemberPriceCouponsPerPeriod int32 `json:"memberPriceCouponsPerPeriod"`
}

// FreezeRequest 是冻结会员的请求体。
//
// Reason 必填：冻结是**收走一个人已经在享的权益**（争议、风控），事后只有这一句话答得上来
// 「为什么这个人三月中旬突然不是会员了」。与入库作废必填原因同一条理由。
type FreezeRequest struct {
	Reason string `json:"reason"`
}

// RevokeRequest 是撤销会员的请求体。
//
// 与冻结分开：**撤销不可逆**（status 回不去，库上的状态机也只允许从 active/frozen 到
// revoked），而冻结是可以解冻的。合用一个接口会让一次手滑变成一次永久撤销。
type RevokeRequest struct {
	Reason string `json:"reason"`
}

// AdjustRequest 是后台直接改会员有效期的请求体。
//
// 这是本域破坏力最大的一个接口：它**没有任何订单、支付或流水跟着发生**，只是把一个人的
// expire_at 挪一下。所以 Reason 必填、备注可选但建议填，且每一次调用都写平台审计（方案
// §11.6「影响用户资产归属的人工操作」）。
//
// 用 ExpireAt 而不是「延长 N 天」：后台要处理的是「这个人应该在 2027-01-01 到期」这类具体
// 结论（对账、客服工单里给的就是一个日期），而「延长 N 天」要求操作的人自己先算一遍，算错的
// 那次没有任何东西能发现。延长语义由前端算完再发过来。
type AdjustRequest struct {
	ExpireAt time.Time `json:"expireAt"`
	Reason   string    `json:"reason"`
	Remark   string    `json:"remark"`
}

// GrantRequest 是后台**直接开通一个会员**的请求体。
//
// 它没有订单号，也没有金额：这是客服补偿、线下活动、渠道争议的出口，钱不在系统里过。它与
// AdjustRequest 一样是本域破坏力最大的接口之一——区别只在于一个是白送一段权益，一个是把已有的
// 那段往后挪。
type GrantRequest struct {
	UserID string `json:"userId"`
	PlanID string `json:"planId"`
	// ExpireAt 留空时按套餐自带时长算（period × periodCount）。给了就用给的：补偿场景常是
	// 「送到年底」这类具体结论，那不是「延长 N 天」能表达的。
	//
	// 与 dto.AdjustRequest 上那个字段同一条写法：类型是 time.Time，所以**没有一条「时间格式
	// 不对」的错误**——写不成时间的值在 controller 解请求体时就已经失败了。
	ExpireAt time.Time `json:"expireAt"`
	// StoreID 是**归属门店**，可留空（见 model.Membership.StoreID）。
	//
	// 它不参与任何权益判定与金额计算，只是「这个人是谁拉来的」这条留痕。后台能改的是**新开通**
	// 那一次的归属；已有会员一律拒绝开通，所以后台也就改不了老会员的归属——「会员过期后重新
	// 开通可变归属」那个例外只属于店铺码活动与在门店买咖啡这两条成交路径。
	StoreID string `json:"storeId"`
	// Reason 必填：这是白送钱，事后只有这一句话答得上来「为什么给他开」。
	Reason string `json:"reason"`
	Remark string `json:"remark"`
	// RequestID 是这次点击的**幂等键**，必填。
	//
	// 客户端在弹窗打开时生成一次、提交时带上；同一个 requestId 重发会拿回同一条会员而不是
	// 409。不能由服务端从 X-Request-Id 取——那是**追踪头**，每个请求都不同。
	RequestID string `json:"requestId"`
}

// AutoRenewRequest 是小程序开关自动续费的请求体。
//
// 它**只能关掉现在开着的那一个**，这也是唯一一个用户自己能改的会员字段。开通自动续费走的是
// 签约那条路（要代扣协议），不是这个接口——见 service 里那段说明。
type AutoRenewRequest struct {
	Enabled bool `json:"enabled"`
}

// MiniappPlanResponse 是小程序看到的套餐形状。
//
// 与 PlanResponse 的差别只有一处，但那一处是必须的：**没有 WechatPlanID，也没有 Status、
// SortOrder、CreatedAt/UpdatedAt**。签约模板 ID 是商户平台的内部标识，透到 C 端没有任何用处，
// 但它会出现在小程序的网络面板里——一个能拿去查我们商户配置的标识不该下发到客户端。
// 其余几个字段是后台的运营属性（草稿态、排序、创建时间），C 端也不该看见。
//
// 分成两个结构而不是给 PlanResponse 加 `json:"-"`：加 omitempty 那种做法改一次后台字段就会
// 悄悄影响 C 端，而这两个契约的**变更节奏完全不同**（后台跟着运营走，C 端跟着发版走）。
type MiniappPlanResponse struct {
	ID          string   `json:"id"`
	Code        string   `json:"code"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Benefits    []string `json:"benefits"`
	PriceCents  int64    `json:"priceCents"`
	Period      string   `json:"period"`
	PeriodCount int32    `json:"periodCount"`
	// AutoRenew 是给 C 端看的「这是连续包月，会到期自动续费」——用户点购买之前要知道这件事。
	AutoRenew bool `json:"autoRenew"`
	// MemberPriceMode 决定购买页上那句权益说明怎么说：「买完自动享会员价」还是
	// 「每期送 N 张会员价券」。两种说法对应两种完全不同的用户预期，不能在客户端猜。
	MemberPriceMode             string `json:"memberPriceMode"`
	MemberPriceCouponsPerPeriod int32  `json:"memberPriceCouponsPerPeriod"`
}

File: admin-web/src/services/membership.ts
import { request } from '@umijs/max';
import type { PageQuery, PageResult } from './pagination';

/**
 * 会员域的接口。字段名与后端 dto（membership-service/internal/dto）的 json tag 逐字对应，
 * 同时也是页面上的 dataIndex——改一处要同批改三处（tag、这里、ProTable 的列），
 * 否则 TypeScript 不报错，只是整列空白。
 *
 * 后端那个包的文件头点名了这一份与小程序会员中心的字段名，说的就是这条约定。
 *
 * 边界先说清楚：**本域只有两样东西**——「卖了什么」（套餐）与「谁在会员中」（会员资格）。
 * 会员是**在别处成交、在这里生效**：买会员那张单在 order-service，会员价的券在
 * coupon-service，代扣签约在微信。所以这个文件里没有订单、没有发券、没有签约，也别加。
 *
 * 连续包月订阅（`membership_subscriptions`）**不在这一份里**，它有自己的
 * services/subscription.ts：那一页是可看的（列表 / 统计 / 详情 + 一个「取消」），但同样
 * **没有创建**——订阅只能由小程序端签约产生。两件事的权限码也分开了（订阅取消是
 * membership:manage，不是 adjust）。
 */

// ——— 枚举 ———
// 取值来自 membership-service model 的常量，也就是 migrations/membership 的 CHECK 约束。
// 接口回的就是库里那些英文码。**改枚举必须同时改迁移和这里**，加了新码而这里没登记，
// 界面上就退回显示原始码。文案表在 services/membershipLabels.ts。

/**
 * membership_plans.status：这个套餐现在卖不卖。
 *
 * **新建的套餐一律是 draft**（CreatePlanRequest 里根本没有 status 字段），要卖得再调一次
 * 上下架接口。三个取值之间没有状态机，任意跳转都行。
 */
export type PlanStatus = 'draft' | 'active' | 'disabled';

/** 每期时长的单位。与 periodCount 一起读才对：「month/1」是连续包月，「year/1」是年度会员。 */
export type PlanPeriod = 'month' | 'year';

/**
 * membership_plans.member_price_mode：会员价从哪来。
 *
 * auto 是「会员本人自动享」，coupon 是「靠会员价体验券」——后者要配券模板与每期张数。
 * 它决定本域那句最要紧的话怎么说：是「你已享会员价」还是「你有 N 张会员价券」。
 */
export type MemberPriceMode = 'auto' | 'coupon';

/**
 * memberships.status：一个人现在的会员资格处在哪一步。
 *
 * frozen 是可逆的（解冻回 active），revoked **不可逆**——所以两者的按钮再像也不能合并。
 * expired 不由后台产生：它只由到期扫描写，后台没有「标记为过期」这个动作。
 */
export type MembershipStatus = 'active' | 'frozen' | 'expired' | 'revoked';

/**
 * membership_changes.change_type：会员身上发生过的一件事。
 *
 * 十四个取值里有几个今天写不出来（subscribe / unsubscribe 要签约链路、refund_adjust 要退款单、
 * expire 要到期扫描），照登是因为迁移里的 CHECK 已经把它们写全了——登记一个今天到不了的码，
 * 好过将来那一格原样显示 `unsubscribe`。
 *
 * charge_failed 与 suspend 来自**已经落地**的代扣链路，两个都写得出来：一期扣款
 * 没扣到写前者（到期日不动），连续失败到上限、停掉自动续费写后者。漏登它们的后果是时间线与
 * 筛选下拉里直接显示英文码。
 */
export type ChangeType =
  | 'activate'
  | 'renew'
  | 'expire'
  | 'freeze'
  | 'unfreeze'
  | 'auto_renew_on'
  | 'auto_renew_off'
  | 'subscribe'
  | 'unsubscribe'
  | 'charge_failed'
  | 'suspend'
  | 'refund_adjust'
  | 'admin_adjust'
  | 'revoke';

/** 这次变动是谁做的。system 是「没有任何人做」——到期扫描那一条就是它。 */
export type MembershipOperatorType = 'user' | 'admin' | 'system' | 'worker';

// ——— 套餐 ———

/**
 * 一个会员套餐。
 *
 * 它是**套餐定义**，不是某个人的会员：这里的 priceCents / period 是「现在买要多少钱、买多久」。
 * 已经买过的人的条款在会员行的快照列上，改这个结构不会动到他们——改价改时长对已购会员
 * 没有影响，这一点是后台敢放开编辑的原因。
 */
export type MembershipPlan = {
  id: string;
  /** 稳定标识，创建后不可改（库上有触发器钉着）。只在新建表单里出现。 */
  code: string;
  name: string;
  description: string;
  /** 「每月 20 次咖啡享会员价」这类卖点，一行一条。 */
  benefits: string[];
  priceCents: number;
  period: PlanPeriod;
  periodCount: number;
  /**
   * 这个**产品**支不支持自动续费（要签微信委托代扣）。
   *
   * 与会员行上的 autoRenew 不是一回事：那个说「这个人现在开着」。套餐今天关掉自动续费，
   * 不等于当初买了它的人不支持。
   */
  autoRenew: boolean;
  /** 微信商户平台的签约模板 ID。autoRenew 为真时必填，为假时必须留空。 */
  wechatPlanId: string;
  memberPriceMode: MemberPriceMode;
  /** 只在 coupon 模式下有值（库上的 CHECK 钉着这对关系）。 */
  memberPriceCouponTemplateId: string;
  memberPriceCouponsPerPeriod: number;
  sortOrder: number;
  status: PlanStatus;
  createdAt: string;
  updatedAt: string;
};

/** 后台套餐列表的筛选条件。 */
export type PlanQuery = PageQuery & {
  /** draft / active / disabled，空表示不筛。**后端不校验这个值**：传错的码返回的是空列表，不是 400。 */
  status?: PlanStatus;
  /** 模糊搜（ILIKE），按名称或编码。 */
  keyword?: string;
};

/**
 * 新建 / 修改套餐的请求体。
 *
 * 新建与修改共用：两者的差别只在「这条记录之前存在吗」，校验是同一套。
 * **没有 status**（上下架是单独的动作，见 updateMembershipPlanStatus）——合在一起的话，
 * 改一次价格描述会把一个正在售的套餐顺手存成草稿，而那一刻可能正有人在下单页上。
 */
export type PlanInput = {
  name: string;
  description: string;
  benefits: string[];
  priceCents: number;
  period: PlanPeriod;
  periodCount: number;
  autoRenew: boolean;
  wechatPlanId: string;
  memberPriceMode: MemberPriceMode;
  memberPriceCouponTemplateId: string;
  memberPriceCouponsPerPeriod: number;
  sortOrder: number;
};

/** 新建套餐的请求体：在 PlanInput 之上多一个 Code（也只有新建给得了它）。 */
export type CreatePlanInput = PlanInput & { code: string };

export async function listMembershipPlans(params?: PlanQuery) {
  return request<PageResult<MembershipPlan>>('/api/v1/admin/membership-plans', { params });
}

export async function getMembershipPlan(id: string) {
  return request<MembershipPlan>(`/api/v1/admin/membership-plans/${id}`);
}

export async function createMembershipPlan(data: CreatePlanInput) {
  return request<MembershipPlan>('/api/v1/admin/membership-plans', { method: 'POST', data });
}

/**
 * 修改套餐。**是整份替换**（后端没有 PATCH 语义），所以调用方必须先把整份详情取出来、
 * 在上面改，再把全量发回来——只发改动过的字段会把没发的字段清空。
 */
export async function updateMembershipPlan(id: string, data: PlanInput) {
  return request<MembershipPlan>(`/api/v1/admin/membership-plans/${id}`, { method: 'PUT', data });
}

/**
 * 上下架 / 转草稿。与「保存表单」分开，理由同抽奖那边：那个是整份换，这个只是开关。
 *
 * 下架**不影响已购会员**：会员行上存着成交快照，套餐不卖了，买过的人拿到的还是当初那些。
 */
export async function updateMembershipPlanStatus(id: string, status: PlanStatus) {
  return request<MembershipPlan>(`/api/v1/admin/membership-plans/${id}/status`, {
    method: 'POST',
    data: { status },
  });
}

// ——— 会员 ———

/** 会员身上发生过的一件事。详情页的时间线，一行一条。 */
export type MembershipChange = {
  id: string;
  changeType: ChangeType;
  /**
   * 状态与到期时间各有一对 from/to，两对都可能**半边为空**：续费只动有效期（状态两个都是
   * active），冻结只动状态。空串表示「没有前一个状态」——开通那一条的 fromStatus 就是空串。
   */
  fromStatus: string;
  toStatus: string;
  fromExpireAt: string | null;
  toExpireAt: string | null;
  /** 到期扫描、后台调整都没有订单，所以这是常态里的常态。 */
  orderId: string | null;
  operatorType: MembershipOperatorType;
  /** 系统与 worker 写的那几条没有操作人。 */
  operatorId: string | null;
  reason: string;
  remark: string;
  occurredAt: string;
};

/**
 * 一条会员资格。
 *
 * 列表与详情同形，差别只在 `changes`：**只有详情填**，列表页是空数组。
 */
export type Membership = {
  id: string;
  /**
   * 用户 ID。**只有它**——会员库没有昵称也没有手机号，接口也不 join 用户服务
   * （跨库，且会员域不该依赖身份域）。再要别的身份信息得另外去问 miniappUsers。
   */
  userId: string;
  /**
   * 套餐编码与名称是**成交当时的快照**，不是现查套餐。套餐后来改了名，这里不跟着变
   * ——用户当初买的就是那个名字。
   */
  planCode: string;
  planName: string;
  /** 同样是成交快照，见 MembershipPlan.memberPriceMode。 */
  memberPriceMode: MemberPriceMode;
  status: MembershipStatus;
  startAt: string;
  expireAt: string;
  /**
   * 这个会员此刻到底还作不作数（status=active **且在有效期内**），**后端算好的**。
   *
   * 不要在前端重算：到期扫描没跑完的那一小段时间里，status 还是 active 而 expireAt 已经
   * 过了，只看 status 的前端会显示「会员有效」。
   */
  active: boolean;
  /**
   * **归属门店**：这个人是谁拉来的，不参与任何金额计算（会员价在哪家店用都一样）。
   *
   * 规则：第一次成为会员那一刻固化，续费与升级不覆盖；**过期之后重新开通**才可变——两个变更
   * 渠道是店铺码活动与在门店买咖啡时开会员。空串表示没有归属，不是错误。
   *
   * storeName 是**后端现解出来的**（会员库只存门店 id，名字是商户域的事实），解不出来时是
   * 空串——门店服务抖一下不会让整页打不开。所以显示时按「空串 ⇒ 显示 id 或 —」处理，
   * 别把空串当成「这家店被删了」。
   */
  storeId: string;
  storeName: string;
  autoRenew: boolean;
  /** 用户关掉自动续费的时刻。开着的时候是 null。 */
  autoRenewOffAt: string | null;
  /** 续了几天期。首购不算续期，所以刚开通是 0。 */
  renewalCount: number;
  lastRenewedAt: string | null;
  /** 冻结与撤销的留痕。时间戳是 null 表示没发生过；两个 reason 没发生过时是**空串**。 */
  frozenAt: string | null;
  freezeReason: string;
  revokedAt: string | null;
  revokeReason: string;
  createdAt: string;
  updatedAt: string;
  /** 时间线，只有详情填。列表页是空数组。 */
  changes: MembershipChange[];
};

/** 后台会员列表的筛选条件。 */
export type MembershipQuery = PageQuery & {
  /** 精确匹配（uuid 列）。后台最常见的一次查询就是「这个用户是不是会员」。 */
  userId?: string;
  /** active / frozen / expired / revoked，空表示不筛。传错的码后端返回 400。 */
  status?: MembershipStatus;
  /** 按套餐编码筛。筛的是**成交快照上的** code，不是现查套餐。 */
  planCode?: string;
  /** 到期时间的闭开区间 [expireFrom, expireTo)，必须是完整的 RFC3339（`2026-01-01` 这种短式会被拒）。 */
  expireFrom?: string;
  expireTo?: string;
  /**
   * 三态：true 只看开着自动续费的，false 只看没开的，不给就是不筛。
   *
   * **别用真值判断**：false 是一个有意义的值，把它当成「没填」会悄悄地少筛一半。
   */
  autoRenew?: boolean;
};

export async function listMemberships(params?: MembershipQuery) {
  return request<PageResult<Membership>>('/api/v1/admin/memberships', { params });
}

/** 取一条会员详情。与列表的区别只有一个：这里带 `changes`。 */
export async function getMembership(id: string) {
  return request<Membership>(`/api/v1/admin/memberships/${id}`);
}

/**
 * 后台**直接开通**会员：给一个还不是会员的人开一条会员。
 *
 * 这是会员域唯一的创建入口。别的会员都是「在别处成交、在这里生效」（买会员的订单付款成功
 * 后发事件过来），只有这一条**没有订单**——客服补偿、线下活动、渠道争议走的就是它。
 *
 * 三条要知道的：
 *
 *  - **已有会员一律 409**，不叠加续期。一个人一条会员，叠加等于把「开通」偷偷变成「续期」，
 *    而那件事在会员详情页有专门的按钮（adjustMembershipExpiry）。重试同一次提交不会 409
 *    ——后端按 requestId 认得出那是同一次点击。
 *  - **到期日默认按套餐算，可以改**。补偿场景常是「送到年底」。
 *  - 权限码是 membership:adjust（**直接白送钱**那一枚），不是 manage。
 */
export type GrantMembershipInput = {
  userId: string;
  planId: string;
  /** 到期时刻（RFC3339）。不传就按套餐的 period / periodCount 从当下算。 */
  expireAt?: string;
  /** 归属门店，可空。给了后端会先问一次「这家店在不在」，不在就拒。 */
  storeId?: string;
  /** 开通原因，必填——这是一次人工白送，审计里要能看出为什么。 */
  reason: string;
  remark?: string;
  /**
   * 幂等键，必填，**由前端在弹窗打开时生成一次**（提交时带上，重试沿用同一个）。
   *
   * 不能省：没有它，一次网络抖动后的重发会让操作员看到「这个人已经是会员了」——而那是他
   * 刚刚亲手建的那条，一句彻头彻尾的假警报。
   */
  requestId: string;
};

export async function grantMembership(data: GrantMembershipInput) {
  return request<Membership>('/api/v1/admin/memberships', { method: 'POST', data });
}

// 下面四个是后台对会员的**人工干预**，权限码是 membership:adjust 而不是 membership:manage
// ——它们直接改一个人已经在享的权益，而改套餐只影响接下来卖什么。每一个都要填原因，
// 每一次调用都会写一条平台审计（方案 §11.6「影响用户资产归属的人工操作」）。

/** 冻结：**只能从 active 来**（不是 active 的会员后端会 409）。不改到期时间。 */
export async function freezeMembership(id: string, reason: string) {
  return request<Membership>(`/api/v1/admin/memberships/${id}/freeze`, {
    method: 'POST',
    data: { reason },
  });
}

/** 解冻：**只能从 frozen 来**。解冻后状态回 active，但「冻结原因」会留着（它记的是上一次冻结）。 */
export async function unfreezeMembership(id: string, reason: string) {
  return request<Membership>(`/api/v1/admin/memberships/${id}/unfreeze`, {
    method: 'POST',
    data: { reason },
  });
}

/** 撤销：除已撤销外都能来，**且不可逆**——后端没有「取消撤销」，它也同时关掉自动续费。 */
export async function revokeMembership(id: string, reason: string) {
  return request<Membership>(`/api/v1/admin/memberships/${id}/revoke`, {
    method: 'POST',
    data: { reason },
  });
}

/**
 * 后台直接改有效期。**是本域破坏力最大的一个接口**：没有任何订单、支付或流水跟着发生，
 * 只是把 expireAt 挪一下。
 *
 * 传的是**绝对时刻**而不是「延长 N 天」：客服工单上给的本来就是一个日期，让操作的人自己
 * 先算一遍，算错的那次没有任何东西能发现。延长语义由调用方算完再发过来。
 * 后端还会挡一条：新的到期时间必须晚于开通时间（否则 409）。
 */
export async function adjustMembershipExpiry(
  id: string,
  data: { expireAt: string; reason: string; remark?: string },
) {
  return request<Membership>(`/api/v1/admin/memberships/${id}/expire`, {
    method: 'POST',
    data,
  });
}
